packages/kolibri/urls.js

```diff
@@ -16,15 +16,119 @@ function generateUrl(baseUrl, { url, origin, port } = {}) {
   return urlObject.href;
 }
 
-const urls = {
-  setUp() {
-    // Set urls onto this object for export
-    // This will add functions for every reversible URL
-    // and strings for __staticURL, __mediaURL, and __contentURL
-    // this behaviour is defined in kolibri/core/kolibri_plugin.py
-    // in the url_tag method of the FrontEndCoreAppAssetHook.
-    Object.assign(this, plugin_data.urls);
-  },
+const baseUrls = [
+  '__staticUrl',
+  '__mediaUrl',
+  '__contentUrl',
+  '__zipContentUrl',
+  '__hashiUrl',
+  '__zipContentOrigin',
+  '__zipContentPort',
+];
+
+class UrlResolver {
+  constructor() {
+    this._patterns = plugin_data?.urls?.urls || {};
+    // Pre-generate all URL functions with variations
```

Comment:
I did originally try to implement it that way around - the issue was the inverse mapping wasn't completely straight forward. As an example, these are all valid URL pattern references currently:
```
'kolibri:core:facility_data-list'
'kolibri:core:facility_data_list'
'kolibri:core:facility_data-arbitrary_action_name'
'kolibri:core:facility_data_arbitrary_action_name'
```
My iniitial implementation was to try to let `_getUrlFunction` take either of these and then do appropriate substitutions to check alternatives, but I couldn't devise a way to do the `_` replacement with `-` that didn't mess with underscores that shouldn't be substituted.

The alternative here would be to tighten our URL references to a subset of what was previously allowed (which I am not opposed to), it would just require a few more updates to eliminate the `_` for `-` instances in the code base (I don't think there are any instances of camelCase usage).